plutus-tx/src/PlutusTx/Extensions/Secrets.hs

```diff
@@ -0,0 +1,133 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications  #-}
+module PlutusTx.Extensions.Secrets(
+  Secret
+  , SecretArgument (..)
+  , mkSecret
+  , secretArg
+  , extractSecret
+  , escape_sha2_256
+  , unsafe_escape_secret
+  ) where
+
+import           Control.Monad
+import           Data.Aeson                   as Aeson (FromJSON (..), ToJSON (..), Value (..))
+import           Data.Aeson.Encoding.Internal (string)
+import           PlutusTx.Prelude             as PlutusTx
+import           Prelude                      as Haskell
+
+-- | A secret value. A value of type `Secret a` can't leak onto
+-- the blockchain in plain-text unless you use an unsafe function.
+-- However, a value of type `Secret a` can end up on the blockchain
+-- via one of the escape hatches like `escape_sha2_256`.
+newtype Secret a = MkSecret a
+
+{- Note [Secret abstraction]
+ - A value of type `Secret a` is guaranteed not to leak (without a compiler warning)
+ - because of type abstraction. Intuitively, the `PlutusTx.Extensions.Secrets` module
+ - does not export the `MkSecret` constructor so as long as the client code that
+ - imports `PlutusTx.Extensions.Secrets` does not use some unholy `unsafePerformIO` to
+ - break the Haskell type system the code is guaranteed not to depend on the actual
+ - value of a secret without:
+ -   1. Using a safe function like `escape_sha2_256` or
+ -   2. Using an unsafe function like `unsafe_escape_secret` (which has a compiler warning
+ -      attached to it).
+ -
+ - This intuition can be made formally precise (see Algehed and Bernardy 2019).
+ -
+ - As noted above, `unsafe_escape_secret` breaks the abstraction barrier and we need to export
+ - it for two reasons:
+ -   1. To make the library backwards-compatible. If you have some new hash function you should
+ -      be able to define `escape_my_magic_hash` without having to implement it in this module.
+ -   2. For tests it is sometimes necessary to use `unsafe_escape_secret` to inspect the value
+ -      of a secret.
+ -
+ - See `Plutus.Contracts.SealedBidAuction` or `Plutus.Contract.GameStateMachineWithSecretTypes`
+ - for examples.
+ -}
+
+-- | Secret argments are provided in the endpoint argument types.
+--
+-- This type guarantees that a `SecretArgument a` that is seen by
+-- the endpoint code is a `Secret a` in a way that can not be
+-- bypassed by safe code.
+data SecretArgument a = UserSide a
```

Comment:
I think this should not live in `plutus-tx`, which doesn't know about "endpoints". Either this bit or the whole thing should go to `plutus-contract`, I guess?

Reply:
That makes sense to me. Should we keep it as `PlutusTx.Extension.Secrets` or does a different module name make sense?

Reply:
How about just `Plutus.Contract.Secrets`?